src/internal.rs

```diff
@@ -76,18 +90,71 @@ macro_rules! __impl_internal_bitflags {
                     if !first {
                         f.write_str(" | ")?;
                     }
+
                     first = false;
                     $crate::__private::core::write!(f, "{:#x}", extra_bits)?;
                 }
 
+                // If no flags are set then write an empty hex flag to avoid
+                // writing an empty string. In some contexts, like serialization,
+                // an empty string is preferrable, but it may be unexpected in
+                // others for a format not to produce any output.
+                //
+                // We can remove this `0x0` and remain compatible with `FromStr`,
+                // because an empty string will still parse to an empty set of flags,
+                // just like `0x0` does.
                 if first {
-                    f.write_str("empty")?;
+                    $crate::__private::core::write!(f, "{:#x}", <$T as $crate::__private::Bits>::EMPTY)?;
                 }
 
                 $crate::__private::core::fmt::Result::Ok(())
             }
         }
 
+        // The impl for `FromStr` should parse anything produced by `Display`
+        impl $crate::__private::core::str::FromStr for $InternalBitFlags {
```

Comment:
This format should be documented somewhere with a grammar so users can generate formatted strings themselves and know they're going to parse.